### PR TITLE
Order

### PR DESCRIPTION
**Pull request by Paradoxallist**

### CODE REVIEW

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    @Override
+    public List<OrderDto> findAll(Long userId) {
+        return orderMapper.toDtoList(orderRepository.findByUserId(userId));
+    }
+
+    @Override
+    @Transactional
+    public OrderDto createOrder(Long userId, CreateOrderRequestDto requestDto) {
+        Order newOrder = new Order();
+        newOrder.setTotal(BigDecimal.valueOf(0));
```
Comment (Rommelua) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]: Why 0 here?

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        return orderItemMapper.toDtoList(orderItemRepository.findByOrderId(orderId));
+    }
+
+    @Override
+    public OrderItemDto getOrderItemByOrderIdAndItemId(Long userId, Long orderId, Long itemId) {
+        userComplianceCheck(userId, orderId);
+        List<OrderItem> orderItemList = orderItemRepository.findByOrderId(orderId);
+        Optional<OrderItem> orderItem = orderItemList.stream()
+                .filter(item -> item.getId().equals(itemId))
+                .findFirst();
```
Comment (Rommelua) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]: Too much in-memory checks. Create method on repository level and pass userId, orderId and itemId there.
You can use Criteria API or custom query

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import store.bookshop.model.OrderItem;
+import store.bookshop.repository.BookRepository;
+import store.bookshop.repository.OrderItemRepository;
+import store.bookshop.repository.OrderRepository;
+import store.bookshop.repository.UserRepository;
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]: do not start class with empty line

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        Set<OrderItem> orderItemSet = new LinkedHashSet<>(orderItems);
+        save.setOrderItems(orderItemSet);
+
+        return orderMapper.toDto(save);
+    }
+
+    @Override
+    @Transactional
+    public OrderDto update(Long orderId,
+                           UpdateOrderRequestDto requestDto) {
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]:
```suggestion
    public OrderDto update(Long orderId, UpdateOrderRequestDto requestDto) {
```

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    @Override
+    public List<OrderItemDto> getOrderItems(Long userId, Long orderId) {
+        return orderItemMapper.toDtoList(
+                orderItemRepository.findByUserIdAndOrderId(userId, orderId));
+    }
+
+    @Override
+    public OrderItemDto getOrderItemByOrderIdAndItemId(Long userId,
+                                                       Long orderId,
+                                                       Long itemId) {
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]:
```suggestion
    public OrderItemDto getOrderItemByOrderIdAndItemId(
                                                       Long userId,
                                                       Long orderId,
                                                       Long itemId
   ) {
```

File: src/main/resources/db/changelog/changes/17-create-orders-table.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                  - id
+            foreignKeyConstraints:
+              - baseColumnNames:
+                  - users_id
+                baseTableName: orders
+                constraintName: fk_orders_user
+                referencedColumnNames:
+                  - id
+                referencedTableName: users
+                onDelete: CASCADE
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/17-create-orders-table.yaml]: file should end with empty line

File: src/main/resources/db/changelog/changes/18-create-order-items-table.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                referencedTableName: orders
+                onDelete: CASCADE
+              - baseColumnNames:
+                  - books_id
+                baseTableName: order_items
+                constraintName: fk_order_items_book
+                referencedColumnNames:
+                  - id
+                referencedTableName: books
+                onDelete: CASCADE
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/18-create-order-items-table.yaml]: file should end with empty line

File: src/main/resources/db/changelog/changes/19-insert-orders.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+                  valueNumeric: 200.00
+              - column:
+                  name: order_date
+                  value: '2024-04-10 10:00:00'
+              - column:
+                  name: shipping_address
+                  value: '123 Main St'
+              - column:
+                  name: is_deleted
+                  valueBoolean: false
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/19-insert-orders.yaml]: file should end with empty line

File: src/main/resources/db/changelog/changes/20-insert-order-items.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                  valueNumeric: 2
+              - column:
+                  name: quantity
+                  valueNumeric: 1
+              - column:
+                  name: price
+                  valueNumeric: 100.00
+              - column:
+                  name: is_deleted
+                  valueBoolean: false
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/20-insert-order-items.yaml]: file should end with empty line

File: src/main/java/store/bookshop/controller/OrderController.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+import store.bookshop.dto.order.UpdateOrderRequestDto;
+import store.bookshop.model.User;
+import store.bookshop.service.order.OrderService;
+
+@Tag(name = "Order manager", description = "Endpoint for managing order")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/controller/OrderController.java]: do not start class with empty line

File: src/main/java/store/bookshop/mapper/OrderItemMapper.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import store.bookshop.config.MapperConfig;
+import store.bookshop.dto.order.CreateOrderItemRequestDto;
+import store.bookshop.dto.order.OrderItemDto;
+import store.bookshop.model.OrderItem;
+
+@Mapper(config = MapperConfig.class, uses = BookMapper.class)
+public interface OrderItemMapper {
+
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/mapper/OrderItemMapper.java]: do not start class with empty line

File: src/main/java/store/bookshop/mapper/OrderMapper.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.List;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import store.bookshop.config.MapperConfig;
+import store.bookshop.dto.order.OrderDto;
+import store.bookshop.model.Order;
+
+@Mapper(config = MapperConfig.class, uses = OrderItemMapper.class)
+public interface OrderMapper {
+
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/mapper/OrderMapper.java]: do not start class with empty line

File: src/main/java/store/bookshop/model/Order.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "orders")
+public class Order {
+
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/model/Order.java]: do not start class with empty line

File: src/main/java/store/bookshop/repository/OrderItemRepository.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+import java.util.List;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import store.bookshop.model.OrderItem;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
+
```
Comment (Vasul-Pankiv) [inline src/main/java/store/bookshop/repository/OrderItemRepository.java]: do not start class with empty line

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        newOrder.setTotal(total);
+        newOrder.setOrderDate(LocalDateTime.now());
+        newOrder.setUser(userRepository.getReferenceById(userId));
+        newOrder.setStatus(Order.Status.COMPLETED);
+        newOrder.setShippingAddress(requestDto.getShippingAddress());
+        Order save = orderRepository.save(newOrder);
+
+        for (OrderItem orderItem : orderItems) {
+            orderItemRepository.save(orderItem);
+        }
```
Comment (Rommelua) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]: orderItems must be already saved (cascade)
add them to order before `orderRepository.save(newOrder);`

File: src/main/java/store/bookshop/dto/order/CreateOrderRequestDto.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+import jakarta.validation.constraints.NotNull;
+import java.util.List;
+import lombok.Data;
+
+@Data
+public class CreateOrderRequestDto {
+    @NotNull
+    private String shippingAddress;
+
+    private List<CreateOrderItemRequestDto> orderItems;
```
Comment (Rommelua) [inline src/main/java/store/bookshop/dto/order/CreateOrderRequestDto.java]: Delete this List.
You should take all cartItems from the ShoppingCart and create the Order - it is not FE job.

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java (hunk 15 above)
Comment (Paradoxallist) [reply src/main/java/store/bookshop/service/order/OrderServiceImpl.java:69]: HHH000437: Attempting to save one or more entities that have a non-nullable association with an unsaved transient entity. The unsaved transient entity must be saved in an operation prior to saving these dependent entities.

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+        newOrder.setTotal(total);
+        newOrder.setOrderDate(LocalDateTime.now());
+        newOrder.setUser(userRepository.getReferenceById(userId));
+        newOrder.setStatus(Order.Status.COMPLETED);
+        newOrder.setShippingAddress(requestDto.getShippingAddress());
+        newOrder.setOrderItems(orderItemSet);
+
+        Order save = orderRepository.save(newOrder);
+
+        return orderMapper.toDto(save);
```
Comment (MentorMA) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]: Let's divide this method and make more readability

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+            Long userId,
+            Long orderId,
+            Long itemId
+    ) {
+        Optional<OrderItem> orderItem =
+                orderItemRepository.findByUserIdAndOrderIdAndItemId(userId, orderId, itemId);
+        if (orderItem.isEmpty()) {
+            throw new EntityNotFoundException("Can't find OrderItem by id: " + itemId);
+        }
+        return orderItemMapper.toDto(orderItem.get());
```
Comment (MentorMA) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]: Use better getOrElseThrow

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+        List<OrderItem> orderItems = new ArrayList<>();
+        BigDecimal total = new BigDecimal(0);
+        for (CartItem cartItem : cartItems) {
+            OrderItem orderItem = new OrderItem();
+            orderItem.setOrder(newOrder);
+            Book book = bookRepository.findById(cartItem.getBook().getId()).orElseThrow(
+                    () -> new EntityNotFoundException("Can't find book by id: "
+                            + cartItem.getBook().getId()));
+            BigDecimal price = book.getPrice();
+            BigDecimal quantity = BigDecimal.valueOf(cartItem.getQuantity());
```
Comment (MentorMA) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java]: Why quantity is BigDecimal?

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/store/bookshop/service/order/OrderServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+        BigDecimal total = calculateTotal(orderItems);
+        newOrder.setTotal(total);
+        newOrder.setOrderItems(new LinkedHashSet<>(orderItems));
+
+        Order savedOrder = orderRepository.save(newOrder);
+
+        return orderMapper.toDto(savedOrder);
+    }
+
+    private ShoppingCart getShoppingCart(Long userId) {
```
Comment (okuzan) [inline src/main/java/store/bookshop/service/order/OrderServiceImpl.java:95]: private methods should go after public ones

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by MentorMA — APPROVED** [review]

**Pull request merged**